Firestore/core/src/firebase/firestore/util/async_queue.cc

```diff
@@ -64,13 +68,37 @@ void AsyncQueue::Enqueue(const Operation& operation) {
   EnqueueRelaxed(operation);
 }
 
+void AsyncQueue::EnqueueAndInitializeShutdown(const Operation& operation) {
+  std::lock_guard<std::mutex> lock{shut_down_mutex_};
+  if (is_shutting_down_) {
+    return;
+  }
+  executor_->Execute(Wrap(operation));
+  is_shutting_down_ = true;
+}
+
+void AsyncQueue::EnqueueEvenAfterShutdown(const Operation& operation) {
+  // Still guarding the lock to ensure sequential scheduling.
+  std::lock_guard<std::mutex> lock{shut_down_mutex_};
+  executor_->Execute(Wrap(operation));
+}
+
+bool AsyncQueue::is_shutting_down() const {
+  return is_shutting_down_;
```

Comment:
Should this read be protected by the mutex as well?

Reply:
Thanks for the catch.